onnxruntime/core/providers/openvino/qdq_transformations/qdq_stripping.cc

```diff
@@ -0,0 +1,671 @@
+    reason = SkipReason::Uint16QDQ;
+    return false;
+  }
+
+  // DQs in Double QDQ cases should be kept
+  if (dq_node->InputDefs().at(2)->Name().find("zero_point_convert") != std::string::npos &&
+      !IsQDQSandwichedBetweenSoftmaxAndConvMatMulOps(dq_node))
+    return true;
+
+  if (op_type == "Conv" || op_type == "MatMul") {
+    // Conv and MatMul always keeps int8 DQs except if the DQ is sandwiched between Softmax and Conv/MatMul
+    if (IsQDQSandwichedBetweenSoftmaxAndConvMatMulOps(dq_node)) {
+      reason = SkipReason::SandwichedDQ;
+      return false;
+    } else {
+      return true;
+    }
+  } else if (op_type == "Add") {
+    // Add keeps all DQs except if it has const inits
+    return !IsAnyDQAConstantInitializer(&target_node, src_graph);
+  } else if (op_type == "Mul" || op_type == "Div") {
+    // Keep DQ of Mul and Div only if the target that preceds it is a supported Op in this list and also check if
+    // inputs of Mul and Div have constant initializers. If they do, then don't keep the DQ.
+    return IsPreviousTargetNodeOfDQValid(dq_node, &target_node, src_graph, true);
+  } else {
+    // Keep DQ of an unsupported Op only if the target that preceds it is a supported Op in this list
+    return IsPreviousTargetNodeOfDQValid(dq_node, &target_node, src_graph, false);
+  }
+}
+
+static bool CheckQRuleSet(const NodeUnit& node_unit,
+                          const Node* q_node,
+                          const onnxruntime::GraphViewer& src_graph,
+                          SkipReason& reason) {
+  // If the target node of the NodeUnit following this one is one of the supported Op types, then keep this Q
+  // This Q should also be uint8
+
+  const auto& target_node = node_unit.GetNode();
+  auto op_type = node_unit.OpType();
+
+  // If UInt16 Q, don't keep it
+  if (GetZeroPointDT(q_node) == ONNX_NAMESPACE::TensorProto_DataType_UINT16) {
+    reason = SkipReason::Uint16QDQ;
+    return false;
+  }
+
+  if (op_type == "Conv" || op_type == "MatMul") {
+    // Conv and MatMul keep all Qs except if the target that succeeds it is Add/Mul/Div AND has any const init
+    return IsNextTargetNodeOfQValid(q_node, &target_node, src_graph, {"Add", "Mul", "Div"}, true);
+  } else if (op_type == "Add") {
+    // Add keeps all Qs
+    return true;
+  } else {
+    // Keep Q of an unsupported Op only if the target that succeeds it is a supported Op in this list
+    return IsNextTargetNodeOfQValid(q_node, &target_node, src_graph, {"Conv", "Add", "MatMul"}, false);
+  }
+}
+
+static bool HandleDoubleQDQ(onnxruntime::Graph& dst_graph, const onnxruntime::GraphViewer& src_graph,
+                            const NodeUnit& node_unit, std::set<std::string>& initializers_to_keep) {
+  int node_unit_input_edge_count = node_unit.InputEdgeCount();
+  int node_unit_output_edge_count = [&]() {
+    int count = 0;
+    for (auto it = node_unit.OutputEdgesBegin(); it != node_unit.OutputEdgesEnd(); ++it)
+      count += 1;
+    return count;
+  }();
```

Comment:
@adrianlizarraga NodeUnit should provide OutputEdgeCount() so this is unnecessary